taichi/analysis/gather_statement_usages.cpp

```diff
@@ -0,0 +1,55 @@
+#include "taichi/ir/ir.h"
+#include "taichi/ir/statements.h"
+#include "taichi/ir/transforms.h"
+#include "taichi/ir/visitors.h"
+
+namespace taichi::lang {
+
+class GatherStatementUsages : public BasicStmtVisitor {
+ private:
+  using BasicStmtVisitor::visit;
+
+  // maps a stmt to all its usages <stmt, operand>
+  std::unordered_map<Stmt *, std::vector<std::pair<Stmt *, int>>> stmt_usages_;
```

Comment:
Can we benchmark whether this or `std::unordered_multimap` with `equal_range` iterators are faster?

Reply:
My attempt shows that they run at the same speed :-)